IdentityCore/src/cache/MSIDTokenCache.m

```diff
@@ -24,50 +24,386 @@
+        {
+            // Todo: add macro and handle error
+            return nil;
+        }
+        if (![scopes isSubsetOfOrderedSet:token.scopes])
+        {
+            continue;
+        }
+        
+        [matchedTokens addObject:token];
+    }
+    
+    *outAuthorityFound = foundAuthority;
+    
+    if (matchedTokens.count == 0)
+    {
+        MSID_LOG_INFO(context, @"No matching access token found.");
+        MSID_LOG_INFO_PII(context, @"No matching access token found.");
+        return YES;
+    }
+    
+    if ([matchedTokens[0] isExpired])
+    {
+        MSID_LOG_INFO(context, @"Access token found in cache is already expired.");
+        MSID_LOG_INFO_PII(context, @"Access token found in cache is already expired.");
+        return YES;
+    }
+    
+    *outAccessToken = matchedTokens[0];
+    return YES;
+}
+
+
+- (MSIDToken *)getRTforUser:(MSIDUser *)user
+                             authority:(NSURL *)authority
+                              clientId:(NSString *)clientId
+                               context:(id<MSIDRequestContext>)context
+                                 error:(NSError **)error
 {
+    // Look for new cache, with utid and uid
+    if (user.userIdentifier)
+    {
+        MSIDTokenCacheKey *key = [MSIDTokenCacheKey keyForRefreshTokenWithUserId:user.userIdentifier
+                                                                        clientId:clientId
+                                                                     environment:authority.msidHostWithPortIfNecessary];
+        
+        MSIDToken *token = [_dataSource itemWithKey:key serializer:_jsonSerializer context:context error:error];
+        if (token)
+        {
+            return token;
+        }
+    }
+    
+    // Look for old cache
+    // if there is upn
+    if (user.upn)
+    {
+        MSIDTokenCacheKey *key = [MSIDTokenCacheKey keyWithAuthority:authority clientId:clientId resource:nil upn:user.upn];
+        
+        MSIDToken *token = [_dataSource itemWithKey:key serializer:_keyedArchiverSerialize context:context error:error];
+        
+        if (token)
+        {
+            return token;
+        }
+    }
+    
     return nil;
 }
 
+- (NSArray<MSIDToken *> *)getAllRTsForClientId:(NSString *)clientId
+                                       context:(id<MSIDRequestContext>)context
+                                         error:(NSError **)error
+{
+    if (!clientId)
+    {
+        return nil;
+    }
+    
+    NSMutableArray<MSIDToken *> *allRTs = [NSMutableArray<MSIDToken *> new];
+    
+    // Look at new cache
+    MSIDTokenCacheKey *key = [MSIDTokenCacheKey keyForRefreshTokenWithClientId:clientId];
+    
+    NSArray *newTokens = [_dataSource itemsWithKey:key serializer:_jsonSerializer context:context error:error];
+    if (!newTokens)
+    {
+        return nil;
+    }
+    
+    for (MSIDToken *token in newTokens)
+    {
+        if (token.tokenType == MSIDTokenTypeRefreshToken)
```

Comment:
can we make a query based on the token type? I think new cache had it.

Reply:
it is in kSecAttrType, not in either account or service.
We could however, revisit datasource and try to get an AT item or RT item to utilize this attribute.

Reply:
I remember that Sergey added tokenType field to the key. If that's the case, can we sync up on that?

Reply:
Yeah, there is a [query setObject:[NSNumber numberWithUnsignedInt:item.tokenType] forKey:(id)kSecAttrType] afaik.
it'd be nice to be able to read specific token types from the query itself, so def. soemthing to sync about